nixos-configuration-on-vm.md: Show how to run headless

nh2 opened the pull request:
Fixes #767.

fricklerhandwerk commented on source/tutorials/nixos/nixos-configuration-on-vm.md:
````diff
@@ -180,13 +180,19 @@ ls -R ./result
 ```
 
-Run the virtual machine:
+To run the virtual machine in a new window (requires GUI environment):
````
```suggestion
Run the virtual machine in the current terminal:
```

Let's not show all the options here, that's what reference documentation is for. The purpose of this tutorial (as any tutorial here) is to serve as a first encounter, with pointers to all the details.

I prefer the terminal version as that is or at least appears more lightweight, and showing a the flag here should provide enough of a hint that omitting it may give you a GUI.

fricklerhandwerk commented on source/tutorials/nixos/nixos-configuration-on-vm.md:
````diff
@@ -180,13 +180,19 @@ ls -R ./result
+./result/bin/run-nixos-vm -nographic
+```
+
+You will see the boot process of the virtual machine and end at the `gdm` login screen where you can log in as `alice` with the password `testpw`.
````
I strongly recommend against describing perceptions, as not everyone can even have them. Describe what the machine does instead.

```suggestion
This will show the boot process of the virtual machine and end at the `gdm` login screen where you can log in as `alice` with the password `testpw`.
```

I currently can't check it myself: will that indeed be a terminal prompt? Isn't GDM a graphical tool?

fricklerhandwerk commented on source/tutorials/nixos/nixos-configuration-on-vm.md:
````diff
@@ -180,13 +180,19 @@ ls -R ./result
 ```
 
-This command opens a window that shows the boot process of the virtual machine and ends at the `gdm` login screen where you can log in as `alice` with the password `testpw`.
+To run it in the current terminal instead, pass `-nographic`:
````
Generally, instead of presenting multiple options - which is distracting to first-time users because they'll be busy with following *one* set of instructions, and also doesn't scale well, since where do you draw the line? – we should link to the manual right afterward, such as:

```
Refer to the Nixpkgs manual for a [list of all commmand-line options to `run-nixos-vm`](<URL>).
```

(If that exists. Otherwise there's likely a `--help` flag?)

fricklerhandwerk reviewed (changes requested):
Thanks for pointing this out. I like the general direction of the change, but think it needs some streamlining.

nh2 commented:
@fricklerhandwerk I've updated the PR.

Your comment was right that of course with the terminal, you can't see the configured GDM.

I think it's important to describe both the GUI and non-GUI options, because

* the article as-is just fails with `gtk initialization failed` when invoking `run-nixos-vm` from a non-GUI NixOS environment, which is extremely confusing (and the reason why I opened this PR, as the instructions didn't work out-of-the-box)
* the non-GUI option isn't perfect either, as you need a GUI to see the configured GDM, and there will likely be as many desktop users as headless users
  * further, the `-nographic` solution suffers from terminal corruption because QEMU serial consoles most of the time do not match the size of the terminal you launch them from; I've added a note about this now, so that users don't have the same "what is this rubbish, none of this works!"-feeling as I got initially.

fricklerhandwerk commented:
> Your comment was right that of course with the terminal, you can't see the configured GDM.

How is this addressed in the PR then?

> the article as-is just fails with gtk initialization failed when invoking run-nixos-vm from a non-GUI NixOS 

We could instead add this to the requirements at the beginning.

> the non-GUI option isn't perfect either

All these problems and the hassle required to work around them indicate that the additional information you collected should be put into the NixOS manual. For instance, that you can pass any QEMU flags, and how to handle terminal scrambling should definitely go into the manual.

I think all this is important to know, but not for the first encounter. Therefore, let's put it in the manual.

fricklerhandwerk commented:
Feel free to reopen if there is something particular you still want to address here.
